ENGAGEHF/Dashboard/Vitals/VitalsManager.swift

```diff
@@ -0,0 +1,353 @@
+        return HKQuantitySample(type: quantityType, quantity: hkQuantity, start: startDate, end: endDate)
+    }
+    
+    private func convertToHKCorrelation(_ observation: R4Observation) throws -> HKCorrelation {
+        // For now, only handle Blood Pressure
+        guard let observationType = observation.code.coding?.first?.code?.value?.string,
+              observationType == "85354-9" else {
+            throw VitalsError.invalidObservationType
+        }
+        
+        let (startDate, endDate) = observation.getEffectiveDate()
+        
+        guard let startDate, let endDate else {
+            throw VitalsError.invalidConversion
+        }
+        
+        // Index into the components of the observation for systolic and diastolic measurements
+        guard let components = observation.component else {
+            throw VitalsError.missingField
+        }
+        
+        let systolicComponent = try self.getComponent(components, code: "8480-6")
+        let diastolicComponent = try self.getComponent(components, code: "8462-4")
+        
+        let systolicQuantity = try self.getQuantity(component: systolicComponent)
+        let diastolicQuantity = try self.getQuantity(component: diastolicComponent)
+        
+        let systolicSample = HKQuantitySample(
+            type: HKQuantityType(.bloodPressureSystolic),
+            quantity: systolicQuantity,
+            start: startDate,
+            end: endDate
+        )
+        let diastolicSample = HKQuantitySample(
+            type: HKQuantityType(.bloodPressureDiastolic),
+            quantity: diastolicQuantity,
+            start: startDate,
+            end: endDate
+        )
+        
+        return HKCorrelation(
+            type: HKCorrelationType(.bloodPressure),
+            start: startDate,
+            end: endDate,
+            objects: [systolicSample, diastolicSample]
+        )
+    }
+    
+    
+    private func getComponent(_ components: [ObservationComponent], code: String) throws -> ObservationComponent {
+        guard let component = components.first(
+            where: {
+                $0.code.coding?.contains(
+                    where: {
+                        $0.code?.value?.string == code
+                    }
+                ) ?? false
+            }
+        ) else {
+            throw VitalsError.missingField
+        }
+        
+        return component
+    }
+    
+    // TODO: Combine the two versions of this into a generic function?
+    private func getQuantity(observation: R4Observation) throws -> HKQuantity {
+        guard case let .quantity(fhirQuantity) = observation.value else {
+            throw VitalsError.invalidConversion
+        }
+        
+        guard let sampleQuantity = fhirQuantity.value?.value?.decimal else {
+            throw VitalsError.invalidConversion
+        }
+        
+        let quantity = NSDecimalNumber(decimal: sampleQuantity).doubleValue
+        
+        let units: HKUnit
```

Comment:
`HKUnit` has a mechanism to get a unit form a string, we should use this: https://developer.apple.com/documentation/healthkit/hkunit/1615733-init. 

Unfortunately Apple is very late in adopting Swift interfaces in HealthKit and the function throws an Objective-C (!) exception. This is quite unfortunate as we can't catch errors like this in Swift (https://developer.apple.com/documentation/swift/handling-cocoa-errors-in-swift#Handle-Exceptions-in-Objective-C-Only). Fun times.

Therefore we should add a very small amount of Objective-C code to the code base or use a dependency like this to catch the error properly without crashing the app based on bad input coming from Firebase or any other external data source that we might not fully trust: https://github.com/sindresorhus/ExceptionCatcher/tree/main